HLSL goal statement and core priorities (#504)
microsoft/hlsl-specs: docs/DesignConsiderations.md

HLSL language features should be portable and safe by default. HLSL strives to
be portable across APIs, hardware vendors, and hardware versions. Non-portable
features should be explicitly marked as such in the source representations (e.g.
in an API-specific namespace, or otherwise explicitly denoted).

Enhancing portability and safety also means curbing undefined behavior and not
sacrificing portability or safety features for performance.

### Public and open-source by default

HLSL depends on industry collaboration, and must prioritize open and equitable
collaboration. We must continue evolving our process and designs to be fully
open and to treat all participants equitably.

### Principle of least astonishment

Most HLSL users are C++ users. Acknowledging that aspects of C++ don't map
efficiently to GPUs (looking at you `virtual`, RTTI, and exceptions), we should
strive for alignment with C++ wherever possible and follow the [principle of
least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

For example, adopting C++'s best-match algorithm for overload resolution aligns
behavior with C++ user expectations.

### We do not exist in a vacuum

> Many of the problems we're solving are not unique to HLSL. We should always look

Reviewer: Use stronger language.
```suggestion
Many of the problems we're solving are not unique to HLSL. We will look
```

to other languages, tools, and ecosystems as we consider how to evolve our own.

### Design for users

HLSL exists to serve users. Consider the experience of users and all the ways
HLSL can empower them to be more productive and creative. HLSL inherits a lot of
sharp edges both from its history and C++; we strive to reduce those cases.

### The Zen of Python is pretty great

While not all of Python's design decisions are applicable to HLSL many of their
design principles apply broadly to programming language design.

[PEP 20 - The Zen of Python](https://peps.python.org/pep-0020/) has a bunch of
deeply relevant pithy sayings which are a great set of guidelines to help shape
programming language design, and we should embrace the wisdom of others.

> Beautiful is better than ugly.
> Explicit is better than implicit.
> Simple is better than complex.
> Complex is better than complicated.
> Flat is better than nested.

> > Sparse is better than dense.

Reviewer 2: That may well be, but some implicit behavior is expected. I don't think anyone is suggesting that we remove the ability to provide implicit parameter values. We are proposing changing initialization such that the unspecified values get implicit assignments. It's not a blanket ban. It requires consideration.
